c++/src/capnp/message.h

```diff
@@ -414,6 +421,21 @@ class FlatMessageBuilder: public MessageBuilder {
   bool allocated;
 };
 
+template <typename MR>
```

Comment:
Yeah, I think we should really make this accept an arbitrary struct reader type, otherwise we might be forcing people to make yet another redundant copy into a MallocMessageBuilder.

I think it's OK to require specifically a struct reader, which means that you can use `.totalSize()` instead of `.targetSize()`. Everything else should work, I think.